src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php

```diff
@@ -184,7 +184,15 @@ private function parseDefinition(\DOMElement $service, $file)
                 } elseif ($childService = $factory->getAttribute('service')) {
                     $class = new Reference($childService, ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE, false);
                 } else {
-                    $class = $factory->getAttribute('class');
+                    $class = $factory->getAttribute('class') ?: $definition->getClass();
+                }
+
+                if (null === $class) {
```

Comment:
Sure it's something that could be enhanced then.
Actually, I already had in mind definitions are defined and altered at different places. With the current state of this PR, if the definition class is changed in a compiler pass for instance, the factory class used won't change. 
I assumed it'll be a known limitation that should not challenge this PR, because it only provides syntax sugar over declaring a factory using YAML/XML formats.

Now, if we should consider this, there are other solutions:

1. #20264 could set the class according to the service id from loaders, because it's probably not the role of the `ContainerBuilder` to alter definitions.
1. We can resolve the factory class in `Definition::getFactory()` instead of computing it in `Definition::setFactory()`
1. Or we can indeed try a more advanced solution with a new `Definition::resolve($id)` method as you suggest, but I don't think we should provide this in this PR nor in #20264